### PR TITLE
[perf] Only use arg.to_uppercase() on json_get when argument size is feasible to be a keyword

### LINKED ISSUE #389: Perf `JSON.GET`: `arg.to_uppercase()` on json_get takes 4,87% out of 30,97% of rejson::json_get. Meaning 17% of overhead on JSON.GET command
To recollect ( Linux based only for now and requires perf ) :

```
make build benchmark PROFILE=1 BENCHMARK=json_get_array_of_docs[1]sclr_pass_100_json.yml PERF_CALLGRAPH_MODE=dwarf
```

### Top entries in text form (notice that the overall cpu% of rejson::json_get is 30,97% )
```
 dso: /root/RedisJSON/target/release/rejson.so
#
# Total Lost Samples: 0
#
# Samples: 3K of event 'cycles:pp'
# Event count (approx.): 131423816178
#
# Overhead  Command       Symbol                                                                                           
# ........  ............  .................................................................................................
#
     4.65%  redis-server  [.] core::unicode::unicode_data::conversions::to_upper
     1.52%  redis-server  [.] <jsonpath_lib::select::Selector as jsonpath_lib::parser::NodeVisitor>::visit_token
     1.17%  redis-server  [.] jsonpath_lib::parser::tokenizer::TokenReader::next_token
     1.13%  redis-server  [.] jsonpath_lib::parser::tokenizer::TokenReader::new
     0.73%  redis-server  [.] jsonpath_lib::parser::Parser::paths_dot
     0.68%  redis-server  [.] hashbrown::raw::inner::RawTable<T>::find
     0.56%  redis-server  [.] rejson::RedisModule_OnLoad::__do_command
     0.54%  redis-server  [.] <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
     0.54%  redis-server  [.] alloc::str::<impl str>::to_uppercase
     0.53%  redis-server  [.] jsonpath_lib::parser::tokenizer::TokenReader::peek_token
     0.45%  redis-server  [.] jsonpath_lib::parser::path_reader::PathReader::take_while
     0.42%  redis-server  [.] jsonpath_lib::parser::Parser::array
     0.39%  redis-server  [.] alloc::raw_vec::finish_grow
     0.34%  redis-server  [.] jsonpath_lib::parser::Parser::key
     0.34%  redis-server  [.] core::str::converts::from_utf8
     0.34%  redis-server  [.] jsonpath_lib::select::FilterTerms::collect_next_with_str
     0.34%  redis-server  [.] __rg_dealloc
     0.34%  redis-server  [.] jsonpath_lib::parser::path_reader::PathReader::next_char
     0.31%  redis-server  [.] jsonpath_lib::parser::NodeVisitor::visit
     0.28%  redis-server  [.] jsonpath_lib::parser::Parser::paths
     0.25%  redis-server  [.] jsonpath_lib::parser::Parser::close_token
     0.25%  redis-server  [.] alloc::raw_vec::RawVec<T,A>::reserve
     0.25%  redis-server  [.] alloc::string::String::insert_bytes
     0.23%  redis-server  [.] rejson::redisjson::RedisJSON::to_json
     0.23%  redis-server  [.] <core::iter::adapters::ResultShunt<I,E> as core::iter::traits::iterator::Iterator>::next
     0.23%  redis-server  [.] alloc::string::String::push
```

### Flame Graph detail

image

svg link:
https://s3.amazonaws.com/benchmarks.redislabs/ci.benchmarks/artifacts/RedisJSON/RedisJSON/profile_json_get_array_of_docs%5B1%5Dsclr_pass_100_json_perf%3Arecord_2021-06-12-23-08-04.out.flamegraph.svg

### PR DESCRIPTION
**Pull request by filipecosta90**
Fixes #389, moving from 137919 ops/sec to 148139 ops/sec. 
To check locally:
```
make build benchmark BENCHMARK=json_get_array_of_docs[1]sclr_pass_100_json.yml
```

Overall difference:
```
"test","rps","avg_latency_ms","min_latency_ms","p50_latency_ms","p95_latency_ms","p99_latency_ms","max_latency_ms"\n
"master -- JSON.GET pass-100 array_of_docs[1].sclr","137919.62","0.110","0.016","0.103","0.151","0.159","3.895"\n'
"branch --  JSON.GET pass-100 array_of_docs[1].sclr","148139.38","0.103","0.016","0.095","0.143","0.143","0.503"\n'
```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -175,27 +175,23 @@ fn json_get(ctx: &Context, args: Vec<String>) -> RedisResult {
...
+            newline = args.next_string()?;
+        } else if arg.eq_ignore_ascii_case("SPACE") {
+            space = args.next_string()?;
+        } else if arg.eq_ignore_ascii_case("NOESCAPE") {
+            continue;
+        } else if arg.eq_ignore_ascii_case("FORMAT") {
+            format = Format::from_str(args.next_string()?.as_str())?;
+        } else {
+            paths.push(Path::new(arg));
+        }
```
Comment (gkorland) [inline src/lib.rs:194]:
```suggestion
        match arg {
            arg if arg.eq_ignore_ascii_case("INDENT") => indent = args.next_string()?,
            arg if arg.eq_ignore_ascii_case("NEWLINE") => newline = args.next_string()?,
            arg if arg.eq_ignore_ascii_case("SPACE") => space = args.next_string()?,
            // Silently ignore. Compatibility with ReJSON v1.0 which has this option. See #168 TODO add support
            arg if arg.eq_ignore_ascii_case("NOESCAPE") => continue,
            arg if arg.eq_ignore_ascii_case("FORMAT") => {
                format = Format::from_str(args.next_string()?.as_str())?
            }
            _ => paths.push(Path::new(arg)),
        };
```

Comment (filipecosta90) [reply src/lib.rs:194]: @gkorland removing `if arg.len() > 8 {` makes us go from 150K ops/sec to ~= 145K ops/sec. 
Meaning solely relying on `eq_ignore_ascii_case` is a good improvement, but using the fast condition check + eq_ignore_ascii_case will be better.
I'll commit the suggestion but add the condition check above.

**Comment by filipecosta90** [comment]
## overall impact of the changes included in this PR. 
**Bottom line from 140K to 150K ops/sec.** @gkorland notice that cd5c201b3ca94d5a16b0fff70a4aa3c58d8b6cf2 is outperforming 5d8a41449e5917b484678b457605b6b392b9c861 as expected given it's cheaper to compare length vs doing ascii based comparison even if the first char is already different. 
 
branch | git_sha | test | rps | avg_latency_ms | min_latency_ms | p50_latency_ms | p95_latency_ms | p99_latency_ms | max_latency_ms
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
master | 6803c258716c1d6d3834b7f03799ffd9ed2e1ff5 | JSON.GET pass-100 array_of_docs[1].sclr | 139848.41 | 0.109 | 0.016 | 0.095 | 0.151 | 0.151 | 0.527
perf.json.get | 5d8a41449e5917b484678b457605b6b392b9c861 | JSON.GET pass-100 array_of_docs[1].sclr | 147050.17 | 0.104 | 0.016 | 0.095 | 0.143 | 0.151 | 0.599
perf.json.get | cd5c201b3ca94d5a16b0fff70a4aa3c58d8b6cf2 | JSON.GET pass-100 array_of_docs[1].sclr | 149351.81 | 0.102 | 0.016 | 0.095 | 0.135 | 0.143 | 0.543

File: src/lib.rs [hunk 2]
```diff
@@ -175,26 +195,19 @@ fn json_get(ctx: &Context, args: Vec<String>) -> RedisResult {
...
-            "FORMAT" => {
-                format = Format::from_str(args.next_string()?.as_str())?;
-            }
-            _ => {
-                paths.push(Path::new(arg));
+        match arg {
+            // fast way to consider arg a path by using the max length of all possible subcommands
+            // See #390 for the comparison of this function with/without this optimization
+            arg if arg.len() > JSONGET_SUBCOMMANDS_MAXSTRLEN => paths.push(Path::new(arg)),
+            arg if arg.eq_ignore_ascii_case("INDENT") => indent = args.next_string()?,
```
Comment (gkorland) [inline src/lib.rs:202]: @filipecosta90 I think you should use constants here and not use "INDENT"

**Review by gkorland — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 3]
```diff
@@ -25,8 +25,34 @@ mod redisjson;
 // use crate::readies_wd40::{BB, _BB, getenv};
 
 const JSON_ROOT_PATH: &'static str = "$";
+const CMD_ARG_NOESCAPE: &'static str = "NOESCAPE";
```
Comment (filipecosta90) [inline src/lib.rs:28]: remove static

**Review by gkorland — APPROVED** [review]

**Pull request merged**